From ada95c3f1ecaad5b74024e159f37145b7a0a4879 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:05:51 -0400
Subject: [virt] kvm/ppc: book3s hv - Implement H_CONFER

Message-id: <1410545655-205645-123-git-send-email-dzickus@redhat.com>
Patchwork-id: 94135
O-Subject: [RHEL7 PATCH 122/626] KVM: PPC: Book3S HV: Implement H_CONFER
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 42d7604d0c5a24f448983424eb19e3845cb031ba
Author: Paul Mackerras <paulus@samba.org>
Date:   Fri Sep 6 13:23:21 2013 +1000

    KVM: PPC: Book3S HV: Implement H_CONFER

    The H_CONFER hypercall is used when a guest vcpu is spinning on a lock
    held by another vcpu which has been preempted, and the spinning vcpu
    wishes to give its timeslice to the lock holder.  We implement this
    in the straightforward way using kvm_vcpu_yield_to().

    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_hv.c b/arch/powerpc/kvm/book3s_hv.c
index dfdd274..d2bb0ca 100644
--- a/arch/powerpc/kvm/book3s_hv.c
+++ b/arch/powerpc/kvm/book3s_hv.c
@@ -538,6 +538,15 @@ int kvmppc_pseries_do_hcall(struct kvm_vcpu *vcpu)
   }
   break;
  case H_CONFER:
+  target = kvmppc_get_gpr(vcpu, 4);
+  if (target == -1)
+   break;
+  tvcpu = kvmppc_find_vcpu(vcpu->kvm, target);
+  if (!tvcpu) {
+   ret = H_PARAMETER;
+   break;
+  }
+  kvm_vcpu_yield_to(tvcpu);
   break;
  case H_REGISTER_VPA:
   ret = do_h_register_vpa(vcpu, kvmppc_get_gpr(vcpu, 4),
-- 
1.7.1